### ISSUE TITLE
computeProposals and evalPerPatch functional, but training fails

### ISSUE THREAD

**Issue by mbuckler** [facebookresearch/deepmask#39]
Hello,

I've installed dependencies and deepmask as per the README instructions, and have be able to successfully run computeProposals on an image, as well as use evalPerPatch to evaluate the pretrained networks on the COCO dataset. The difficulty that I encounter is that when attempting to run training, I receive the following error:

```
root@735ebbd730d6:~/deepmask-repo# th train.lua
-- ignore option rundir	
-- ignore option dm	
-- ignore option reload	
-- ignore option gpu	
-- ignore option datadir	
| running in directory /root/deepmask-repo/exps/deepmask/exp	
/root/torch/install/bin/luajit: /root/torch/install/share/lua/5.1/inn/utils.lua:72: assertion failed!
stack traceback:
	[C]: in function 'assert'
	/root/torch/install/share/lua/5.1/inn/utils.lua:72: in function 'callback'
	/root/torch/install/share/lua/5.1/nn/Module.lua:388: in function 'replace'
	/root/torch/install/share/lua/5.1/nn/Module.lua:391: in function 'replace'
	/root/torch/install/share/lua/5.1/inn/utils.lua:68: in function 'BNtoFixed'
	/root/deepmask-repo/DeepMask.lua:59: in function 'createTrunk'
	/root/deepmask-repo/DeepMask.lua:28: in function '__init'
	/root/torch/install/share/lua/5.1/torch/init.lua:91: in function </root/torch/install/share/lua/5.1/torch/init.lua:87>
	[C]: in function 'DeepMask'
	train.lua:95: in main chunk
	[C]: in function 'dofile'
	/root/torch/install/lib/luarocks/rocks/trepl/scm-1/bin/th:145: in main chunk
	[C]: at 0x00406670
```

In an effort to understand why this might be occurring I found the failing assertion in /root/torch/install/share/lua/5.1/inn/utils.lua on line 72. A subsection of this file is shown below, where the last line with the assertion is line 72:
```

function utils.BNtoFixed(net, ip)
  local net = net:replace(function(x)
      if torch.typename(x):find'BatchNormalization' then
        local no = x.running_mean:numel()
        local y = inn.ConstAffine(no, ip):type(x.running_mean:type())
        assert(x.running_var and not x.running_std)
```

Unfortunately I'm not able to find other information within the issues, and am not yet skilled enough in torch to be able to interpret this error/failed assertion. Is there a known solution to this issue?

Thanks

**Comment by poliveirap** [comment]
Hi, I think you are not using the most up-to-date version. 
Could you pull an retry? The error should go away

**Comment by mbuckler** [comment]
Hi poliveirap, 

Unfortunately this is with the most up to date version. I've attempted to pull and verified that I am on the most recent commit. Do you know of any problems like this that used to happen before a bug fix? The solution to those problems may shed light on why I am experiencing this error.

```
root@fad18523225f:~/deepmask-repo# git pull
Already up-to-date.
root@fad18523225f:~/deepmask-repo# git log -1
commit e6a4169bf8787c1b5421d97b82e50da8a9707547
Merge: ebc7bc5 8e45b39
Author: Soumith Chintala <soumith@gmail.com>
Date:   Tue Aug 30 18:47:53 2016 -0400

    Merge pull request #12 from kevintpeng/patch-1
    
    remove duplicate dependency in README (2x cudnn)
```

**Comment by mbuckler** [comment]
Hi @poliveirap ,

After some more study it looks like the issue is within DeepMask.lua, specifically line 59 which is intended to remove the batch normalization layers from the pre-trained resnet. I'm not yet sure what do with this information, but I have also noticed that the DeepMask.lua is still on its initial commit. If you say that I have an old version of the code then it is possible that there may be a commit that you folks have locally which hasn't yet been pushed. Sorry for the bother, but thanks for the help!

**Comment by poliveirap** [comment]
Hi @mbuckler.
No, thats not the problem, its normal to not have any modification to deepmask.lua on the last push.

I also try to just clone a new deepmask repo and it worked normal in my side.
It's really difficult to help since I cannot reproduce the error :(...

It seem that the BNtoFixed() from utils.lua from imagine-nn package is not finding certain variables in the batchnorm layer of the resnet (which is weird, because it works here).

What is the version of resnet-50 you are using?
For me, md5sum resnet-50.t7 gives
dee966fbac8b0594903494029046d1af

**Comment by mbuckler** [comment]
Hi @poliveirap ,

Thank you very much, with your help I was able to get training to run! Granted I haven't had time to let training complete yet, but I see my GPU utilization and memory usage jump up, and see epoch messages. I needed two changes which are both different from what the installation instructions currently say which is a bit odd. At any rate, here are the details so future people will know whats up and perhaps it may be worth suggesting changes to the installation instructions.

1) I did have that version of resnet-50 with the same md5sum hash that you have, but it turns out that doesn't work for me. On a whim I downloaded the trained resnet-50 from here instead: https://github.com/facebook/fb.resnet.torch/tree/master/pretrained and it turns out that it works! If you are using the resnet model referenced in the current deepmask readme (the one that failed for me) then that is strange since it doesn't work on my end. Either way, this is probably enough info for devs/other folks installing.

2) I encountered this issue and solved it by using luarocks install nn. https://github.com/facebookresearch/deepmask/issues/22. Currently the installation readme for deepmask asks folks to install nnx and inn, but not nn. It is worth adding to the readme that nn should be installed as well.

Thanks again for the help in questioning the actual resnet-50 model. I'll upload a dockerfile so that other folks can get started instantly
